modules/plottinggl/src/plotters/scatterplotgl.cpp

```diff
@@ -604,52 +491,96 @@ void ScatterPlotGL::objectPicked(PickingEvent* p) {
-
+    if (properties_.tooltip_ && (p->getHoverState() == PickingHoverState::Move ||
+                                 p->getHoverState() == PickingHoverState::Enter)) {
         tooltipCallback_.invoke(p, id);
-
     } else if (p->getHoverState() == PickingHoverState::Exit) {
+        // unset tooltip at all times in case one was set prior disabling tooltips
         p->setToolTip("");
     }
 
-    if (properties_.hovering_.get()) {
+    if (properties_.showHighlighted_) {
         if (p->getHoverState() == PickingHoverState::Enter ||
             (p->getGlobalPickingId() != p->getPreviousGlobalPickingId())) {
-            highlighted_.clear();
-            highlighted_.add(id);
-            highlightChangedCallback_.invoke(highlighted_);
-            highlightDirty_ = true;
+            highlightChangedCallback_.invoke(BitSet(id));
         } else if (p->getHoverState() == PickingHoverState::Exit) {
-            highlighted_.clear();
-            highlightChangedCallback_.invoke(highlighted_);
+            highlightChangedCallback_.invoke(BitSet());
         }
     }
 
+    const double movedDistance =
+        glm::length(p->getDeltaPressedPosition() * dvec2(p->getCanvasSize()));
     if ((p->getPressState() == PickingPressState::Release) &&
-        (p->getPressItem() == PickingPressItem::Primary) &&
+        (p->getPressItem() == PickingPressItem::Primary) && (movedDistance < 3.0) &&
         (p->getCurrentGlobalPickingId() == p->getPressedGlobalPickingId())) {
-        ensureSelectAndFilterSizes();
-        selected_[id] = !selected_[id];
-        selectedIndicesGLDirty_ = true;
-
-        // selection changed, inform processor
-        selectionChangedCallback_.invoke(selected_);
+        if (p->modifiers().contains(KeyModifier::Control)) {
+            // additive selection: add/remove current id
+            BitSet selected(selected_);
+            selected.flip(id);
+            selectionChangedCallback_.invoke(selected);
+        } else {
+            selectionChangedCallback_.invoke(BitSet(id));
+        }
+        // reset box selection handler since the data point was selected here
+        boxSelectionHandler_.reset();
+        p->setUsed(true);
     }
-    p->setUsed(true);
 }
 
 uint32_t ScatterPlotGL::getGlobalPickId(uint32_t localIndex) const {
     return static_cast<uint32_t>(picking_.getPickingId(localIndex));
 }
 
-void ScatterPlotGL::ensureSelectAndFilterSizes() {
-    if (xAxis_->getSize() != selected_.size() || xAxis_->getSize() != filtered_.size()) {
-        selected_.resize(xAxis_->getSize(), false);
-        filtered_.resize(xAxis_->getSize(), false);
-        selectedIndicesGLDirty_ = true;
-        filteringDirty_ = true;
+void ScatterPlotGL::partitionData() {
+    const auto numDataPoints = static_cast<std::uint32_t>(points_.xCoord->getSize());
+    auto seq = util::make_sequence<std::uint32_t>(0, numDataPoints);
+
+    auto& indices = points_.indices.getEditableRAMRepresentation()->getDataContainer();
```

Comment:
Why do you always reset the list at the start? seems like this will have to do a lot more work than needed.

Reply:
in most cases on only has to move a few point around, like the hovered one etc.